### PR TITLE
Add DataMirror API for currentInstancePorts.

### PR DESCRIPTION
**Pull request by mikeurbach**
### Contributor Checklist

- [x] Did you add Scaladoc to every public function/method?
- [x] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [ ] Did you add appropriate documentation in `docs/src`?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

- Feature (or new API)

#### Desired Merge Strategy

- Squash: The PR will be squashed and merged (choose this if you have no preference).

#### Release Notes

This allows users to query an Instance without it being closed or forcing it to close, which can be useful in atModuleBodyEnd and other scenarios.

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels? (Select the most appropriate one based on the "Type of Improvement")
- [ ] Did you mark the proper milestone (Bug fix: `3.6.x`, `5.x`, or `6.x` depending on impact, API modification or big change: `7.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you do one of the following when ready to merge:
  - [ ] Squash: You/ the contributor `Enable auto-merge (squash)`, clean up the commit message, and label with `Please Merge`.
  - [ ] Merge: Ensure that contributor has cleaned up their commit history, then merge with `Create a merge commit`.

### CODE REVIEW

File: src/test/scala/chiselTests/reflect/DataMirrorSpec.scala [hunk 1]
```diff
@@ -266,6 +266,22 @@ class DataMirrorSpec extends ChiselFlatSpec {
...
+      @public val in = IO(Input(Bool()))
+      @public val out = IO(Output(Bool()))
+    }
+
+    ChiselStage.emitCHIRRTL(new RawModule {
+      val foo = Instantiate(new Foo)
+
+      val ports = DataMirror.internal.currentInstancePorts(foo)
+
+      ports should be(Seq(foo.in, foo.out))
```
Comment (jackkoenig) [inline src/test/scala/chiselTests/reflect/DataMirrorSpec.scala:281]: You should also test a `Module(new Foo)`, if possible, also see if you can box the anonymous `RawModule` in an `Instance` and check its ports (if so, give it a port).

File: core/src/main/scala/chisel3/reflect/DataMirror.scala [hunk 2]
```diff
@@ -265,6 +266,24 @@ object DataMirror {
...
+      * data structures to extract the module's IOs. For this reason, it is generally not safe, and users should prefer
+      * `DataMirror.modulePorts`, but this method may be used for certain use cases that want the current list of
+      * ports before the module is closed.
+      *
+      * @param target Instance to get IOs from
+      */
+    def currentInstancePorts(target: Instance[BaseModule]): Seq[Data] = {
+      target.underlying match {
+        case Clone(clone: ModuleClone[_]) =>
+          clone.getPorts.elements.map(_._2).toSeq
```
Comment (jackkoenig) [inline core/src/main/scala/chisel3/reflect/DataMirror.scala:282]:
```suggestion
          clone.getPorts.elementsIterator.toSeq
```
This saves an iteration and is inline with my medium-term goal of changing `elements` to be a `def` which will make calling it expensive.

**Review by jackkoenig — APPROVED** [review]
LGTM but see comments

File: src/test/scala/chiselTests/reflect/DataMirrorSpec.scala (hunk 1 above)
Comment (mikeurbach) [reply src/test/scala/chiselTests/reflect/DataMirrorSpec.scala]: Thanks for the pointers, added in [6b378ea](https://github.com/chipsalliance/chisel/pull/4362/commits/6b378ea9f8f638b8b8b09e671b21763354869f3f)

File: core/src/main/scala/chisel3/reflect/DataMirror.scala (hunk 2 above)
Comment (mikeurbach) [reply core/src/main/scala/chisel3/reflect/DataMirror.scala:282]: Got it, will do. I always forget I can use these APIs from within Chisel.

**Pull request merged**